docs/proposals/24-shoot-oidc-issuer.md

```diff
@@ -0,0 +1,128 @@
+
+### Goals
+ - Stay backwards compatible with the current configuration options. Existing integrations should continue to work.
+ - Provide a secure managed solution for serving the OIDC discovery documents of a shoot cluster.
+ - Improve integration capabilities for clusters that reside in a fenced environment.
+### Non-Goals
+ - Establish trust relations between clusters. This GEP is about simplifying the prerequisites for doing that.
+
+## Proposal
+
+This proposal consists of three parts that together will achieve the mentioned goals.
+
+### Static validation
+
+Static validation rule will be introduced enforcing that `shoot.Spec.Kubernetes.KubeAPIServer.ServiceAccountConfig.Issuer` is not set when the shoot is annotated with `gardener.cloud/managed-issuer: true`. The rule will also assure that once present the annotation cannot be removed from a shoot.
+
+```go
+if oldShoot.Annotations["gardener.cloud/managed-issuer"] == "true" {
+    // ensure that the new shoot also has this annotation
+    // ensure that newShoot.Spec.Kubernetes.KubeAPIServer.ServiceAccountConfig.Issuer is not set
+    newShoot.Annotations["gardener.cloud/managed-issuer"] = "true"
+} else if newShoot.Annotations["gardener.cloud/managed-issuer"] == "true" {
+    // ensure that newShoot.Spec.Kubernetes.KubeAPIServer.ServiceAccountConfig.Issuer is not set
+}
+```
+
+### Gardenlet
+
+Once annotated shoots will be easily identifiable by the `gardenlet`. If a shoot requires a managed issuer then the `gardenlet` will configure the `--service-account-issuer` and `--service-account-jwks-uri` flags for the shoot's [kube-apiserver](https://kubernetes.io/docs/reference/command-line-tools-reference/kube-apiserver/) deployment during reconciliation and set their values to `https://<central.domain.name>/projects/<project-name>/shoots/<shoot-uid>/issuer` and `https://<central.domain.name>/projects/<project-name>/shoots/<shoot-uid>/issuer/jwks` respectively (the central domain name should be configurable). At the end of a shoot reconciliation flow, the `gardenlet` will fetch the OIDC metadata discovery documents from a shoot's `kube-apiserver` and sync them back to the Garden cluster in the form of a configmap following the naming convention `projectname--shootuid` in a special purposed namespace called `gardener-shoot-issuer`(name should be configurable). An additional advertised address will be added to the shoot status so that the issuer of the cluster is easily discoverable.
+
+```yaml
+status:
+  advertisedAddresses:
+    ...
+    - name: issuer
```

Comment:
```suggestion
    - name: serviceAccountIssuer # or service-account-issuer
```

or 

```suggestion
    - name: OIDCIssuer # or oidc-issuer
```

This will leave some space for future use cases that might also have `issuer` in their name.

Reply:
Yes, a valid point. I will address that.